File: leetcode/array/0001_two_sum/solution.cpp
# include <vector>

using namespace std;

// 自己写的暴力解法
class Solution_OWN {
    public:
        vector<int> twoSum(vector<int> &nums, int target)
        {
            for(int i = 0; i < nums.size(); i++)
            {
                for(int j = i + 1; j < nums.size(); j++)
                {
                    if(nums[i] + nums[j] == target)
                    {
                        return {i, j};
                    }
                }
            }
            return {};
        }
};

// 题解给出的哈希表
# include <unordered_map>

class Solution_Pro{
    public:
        vector<int> twoSum(vector<int> & nums, int target)
        {
            unordered_map<int, int> hastable;
            for(int i = 0; i < nums.size(); i++)
            {
                auto it = hastable.find(target - nums[i]);
                if(it != hastable.end())
                {
                    return {it->second, i};
                }
                hastable[nums[i]] = i;
            }
            return {};
        }
};

File: leetcode/array/0001_two_sum/solution.py
from typing import List

class Solution_OWN:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)
        for i in range(n):
            for j in range(i + 1, n):
                if nums[i] + nums[j] == target:
                    return [i,j]
        return []
    

import hash_table

class Solution_Pro:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashtable = {}
        for i, num in enumerate(nums):
            if target - num in hashtable:
                return [hashtable[target - num], i]
            hashtable[num] = i
        return []
    
